ECS-01 feat: cria cluster ecs, service e task definition (#2)
mentoriaiac/iac-modulo-aws-ecs: README.md

## Infraestrutura
Imagem da infraestrutura do cluster ECS.

![image](./img/ecs-mentoria.png)
## Requirements

| Name | Version |
|------|---------|
| <a name="requirement_terraform"></a> [terraform](#requirement\_terraform) | >= 1.0.0 |
| <a name="aws"></a> [aws](#requirement\aws) | ~> 4.0 |

## Providers

| Name | Version |
|------|---------|
| <a name="provider_aws"></a> [aws](#aws\_aws) | 4.0 |

## Modules

Este modulo permite provisionar cluster ECS.

### O que será criado?
- Cluster ECS.
- Service.
- Task definition.
- Politicas de AutoScaling
- Security Group.
- Load Balance.

### O que é o Amazon ECS?
O Amazon Elastic Container Service (ou ECS, para abreviar) é um serviço totalmente gerenciado de orquestração de contêineres que ajuda a implantar, gerenciar e escalar facilmente aplicações em contêineres.

> Para saber mais sobre AWS ECS acesso o link: https://docs.aws.amazon.com/ecs/index.html 

## Recursos utilizados nesse projeto:

| Name | Type | Description |
|------|------|---------- |
| [aws_ecs_cluster](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecs_cluster) | resource | Cluster para excutar os containers.  |

> | [aws_ecs_service](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecs_service) | resource | Permite executar e manter uma simultaneamente no número especificado de tasks em execução no cluster.  |

Reviewer: > Permite executar e manter uma simultaneamente [...]

Acho que faltou uma palavra aqui?

> | [aws_ecs_task_definition](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecs_task_definition) | resource | É um arquivo de texto `em formato JSON` que descreve um ou mais características dos contêineres. |

Reviewer: ```suggestion
| [aws_ecs_task_definition](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecs_task_definition) | resource | É um arquivo de texto em formato JSON que descreve uma ou mais características dos contêineres. |
```

| [aws_appautoscaling_target](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/appautoscaling_target) | resource | O Auto Scaling é um serviço que permite aumentar ou diminuir a escala das tarefas (tasks). |

> | [aws_appautoscaling_policy](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/appautoscaling_policy) | resource | O Auto Scaling é um serviço que permite aumentar ou diminuir a escala das tarefas (tasks). |

Reviewer: ```suggestion
| [aws_appautoscaling_policy](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/appautoscaling_policy) | resource | O Auto Scaling policy define as regras de escala de um serviço Auto Scaling . |
```

| [aws_iam_role](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role) | resource | IAM é um serviço de gerenciamento de acesso que ajuda você a controlar o acesso aos recursos da AWS. |
| [aws_iam_role_policy_attachment](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_policy_attachment) | resource | IAM é um serviço de gerenciamento de acesso que ajuda você a controlar o acesso aos recursos da AWS. |
| [aws_lb](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb) | resource | Elastic Load Balance distribui automaticamente o tráfego de entrada entre as tarefas (tasks) do serviço no cluster. |

> | [aws_lb_target_group](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lbhttps://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_target_group) | resource | Fornece um grupo de destino que onde o load balance consiga encaminhar requisições. |

Reviewer: ```suggestion
| [aws_lb_target_group](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_target_group) | resource | Fornece um grupo de destino que onde o load balance consiga encaminhar requisições. |
```

| [aws_lb_listener](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_listener) | resource | Porta onde o balanceador de carga está escutando as requisições. |
| [aws_security_group](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/elasticache_security_group) | resource |  Grupo de segurança atua como um firewall virtual. |
| [aws_cloudwatch_log_group](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudwatch_log_group) | resource | CloudWatch responsável por armazenar os logs das task. |

> | [required_providers](https://www.terraform.io/language/expressions/version-constraints) | expression | Sintaxe do Terraform para restrições de versão. |
> | [provider](https://www.terraform.io/language/providers) | provider | São plugins utilizado pelo Terraform para interagir com provedores de nuvem.
> | [locals](https://www.terraform.io/language/values/locals) | input | Um conjunto de valores relacionados `chave=valor` que podem ser declarados em um único bloco. |
> | [variables](https://www.terraform.io/language/values/variables) | input | Serve como parâmetros para que um usuário possam personalizar o comportamento sem editar o código fonte. |

Reviewer: ```suggestion
```

Acho que esses aqui não precisam porque são comuns a todos os módulos.

Reviewer 2: Decidimos que deixaremos essa parte para facilitar o entendimento de quem está começando com Terraform

## Inputs

| Name | Description | Type | Default | Required |
|------|-------------|------|---------|:--------:|

> | cria\_cluster | Define se cluster será criado | `bool` | true | no |

Reviewer: ```suggestion
| create\_cluster | Define se cluster será criado | `bool` | true | no |
```
Eu não sei se a gente já padronizou isso, mas as outras variáveis estão em inglês, então é bom seguir um padrão.

| cluster\_name | Nome do cluster ECS | `string` | n/a | yes |
| container\_insights | Usado para habilitar CloudWatch Container Insights para o cluster | `bool` | true | no |
| delete\_protection | Impede que terraform exclua o load balance | `bool` | false | no |
| region | Região AWS | `string` | n/a | yes |
| service\_name | Nome do service cluster que será criado | `string` | n/a | yes |
| tags | Tags para recurso | `map(string)` | n/a | yes |

> | tags | Tags para recurso | `map(string)` | n/a | yes |

Reviewer: ```suggestion
```

Essa linha acabou duplicada.

| region | Região AWS | `string` | us-west-2 | yes |
| service\_name | Nome do service cluster que será criado | `string` | n/a | yes |
| app\_count | Números de tarefas em execução task definition | `number` | n/a | yes |
| family\_name | Nome para task definition | `string` | n/a | yes |

> | fargate\_cpu | Número de CPUs usados na taskde finition | `number` | n/a | yes |

Reviewer: ```suggestion
| fargate\_cpu | Número de CPUs usados na task definition | `number` | n/a | yes |
```

| fargate\_memory | Quantidade de memória usada pela task definition | `number` | n/a | yes |
| subnet\_ids | ID das subnetes | `list(string)` | n/a | yes |
| vpc\_id | ID da VPC | `string` | n/a | yes |
| app\_port | Porta que será utilizada pela aplicação | `number` | n/a | yes |
| protocol | Protocolo que será utilizado na aplicação `<http, https, tcp>` | `string` | n/a | yes |
| policy\_ssl | Nome da política SSL. Obrigatório se o protocolo for `HTTPS ou TLS` | `string` | null | no |
| certificate\_arn | ARN do certificado de servidor SSL padrão | `string` | null | no |

> | template\_container | Um arquivo json que contém as definições do container | `list(object` | n/a | no |

Reviewer: ```suggestion
| template\_container | Objetos que contém as definições dos containers | `list(object)` | n/a | yes |
```

Isso é uma lista de objects né?

Author: Sim.
